tests/vxlan/test_vxlan_ecmp.py

```diff
@@ -621,13 +662,462 @@ def dump_self_info_and_run_ptf(self, tcname, encap_type, expect_encap_success):
+        # It is one of the arguments to the ptf code.
+        all_t2_ports = list(self.setup[encap_type]['t2_ports'])
+
+        # A distinction in this script between ports and interfaces:
+        # Ports are physical (Ethernet) only.
+        # Interfaces have IP address(Ethernet or PortChannel).
+
+        try:
+            selected_intfs = []
+            # Choose some intfs based on the parameter ecmp_path_count.
+            # when ecmp_path_count == 1, it is non-ecmp. The switching happens between ecmp and non-ecmp.
+            # Otherwise, the switching happens within ecmp only.
+            for i in range(ecmp_path_count):
+                selected_intfs.append(all_intfs[i])
+
+            for intf in selected_intfs:
+                self.setup['duthost'].shell("sudo config interface shutdown {}".format(intf))
+            downed_ports = get_corresponding_ports(selected_intfs, minigraph_facts)
+            self.setup[encap_type]['t2_ports'] = list(set(all_t2_ports) - set(downed_ports))
+            downed_bgp_neighbors = get_downed_bgp_neighbors(selected_intfs, minigraph_facts)
+            pytest_assert(wait_until(300, 30, 0, bgp_established, self.setup['duthost'], downed_bgp_neighbors), "BGP neighbors didn't come up after all interfaces have been brought up.")
+            self.dump_self_info_and_run_ptf("tc12", encap_type, True, packet_count=1000)
+
+            # Reverse the action: bring up the selected_intfs and shutdown others.
+            for intf in selected_intfs:
+                self.setup['duthost'].shell("sudo config interface startup {}".format(intf))
+            # Shutdown other interfaces.
+            remaining_interfaces = list(set(all_intfs) - set(selected_intfs))
+            for intf in remaining_interfaces:
+                self.setup['duthost'].shell("sudo config interface shutdown {}".format(intf))
+            downed_bgp_neighbors = get_downed_bgp_neighbors(remaining_interfaces, minigraph_facts)
+            pytest_assert(wait_until(300, 30, 0, bgp_established, self.setup['duthost'], downed_bgp_neighbors), "BGP neighbors didn't come up after all interfaces have been brought up.")
+            self.setup[encap_type]['t2_ports'] = get_corresponding_ports(selected_intfs, minigraph_facts)
+            self.dump_self_info_and_run_ptf("tc12", encap_type, True, packet_count=1000)
+
+            # recovery. Bring all up, and verify traffic works.
+            for intf in all_intfs:
+                self.setup['duthost'].shell("sudo config interface startup {}".format(intf))
+            # Wait for all bgp is up.
+            pytest_assert(wait_until(300, 30, 0, bgp_established, self.setup['duthost']), "BGP neighbors didn't come up after all interfaces have been brought up.")
+            # Verify traffic flows after recovery.
+            self.setup[encap_type]['t2_ports'] = all_t2_ports
+            self.dump_self_info_and_run_ptf("tc12", encap_type, True, packet_count=1000)
+
+        finally:
```

Comment:
Changing it to exception instead of finally also means we need to re-raise the exception. Added a raise at the end, so the test will be marked for failure if except block is executed.